src/libasr/pass/pass_list_concat.cpp

```diff
@@ -0,0 +1,234 @@
+            al, loc, 0, int_type));
+        do_loop_head.m_end = ASRUtils::EXPR(ASR::make_IntegerBinOp_t(
+            al, loc, ASRUtils::EXPR(ASR::make_ListLen_t(
+                al, loc, arg_exprs[0], int_type, nullptr)
+            ), ASR::binopType::Sub,
+            ASRUtils::EXPR(ASR::make_IntegerConstant_t(
+                al, loc, 1, int_type)
+            ), int_type, nullptr));
+        do_loop_head.m_increment = ASRUtils::EXPR(make_IntegerConstant_t(
+            al, loc, 1, int_type));
+        do_loop_head.loc = loc;
+        Vec<ASR::stmt_t*> loop_body;
+        loop_body.reserve(al, 1);
+        ASR::stmt_t* list_append = ASRUtils::STMT(ASR::make_ListAppend_t(
+            al, loc, arg_expr, ASRUtils::EXPR(ASR::make_ListItem_t(
+                al, loc, arg_exprs[0], do_loop_head.m_v, item_type, nullptr))));
+        loop_body.push_back(al, list_append);
+        list_concat_stmt = ASRUtils::STMT(ASR::make_DoLoop_t(
+            al, loc, do_loop_head, loop_body.p, loop_body.n));
+        Vec<ASR::stmt_t*> list_while_loop = PassUtils::replace_doloop(
+            al, *ASR::down_cast<ASR::DoLoop_t>(list_concat_stmt));
+        for( size_t i = 0; i < list_while_loop.size(); i++ ) {
+            body.push_back(al, list_while_loop[i]);
+        }
+
+        // Copy `right_list` contents
+        do_loop_head.m_end = ASRUtils::EXPR(ASR::make_IntegerBinOp_t(
+            al, loc, ASRUtils::EXPR(ASR::make_ListLen_t(
+                al, loc, arg_exprs[1], int_type, nullptr)
+            ), ASR::binopType::Sub,
+            ASRUtils::EXPR(ASR::make_IntegerConstant_t(
+                al, loc, 1, int_type)
+            ), int_type, nullptr));
+        loop_body.p = nullptr, loop_body.n = 0;
+        loop_body.reserve(al, 1);
+        list_append = ASRUtils::STMT(ASR::make_ListAppend_t(
+            al, loc, arg_expr, ASRUtils::EXPR(ASR::make_ListItem_t(
+                al, loc, arg_exprs[1], do_loop_head.m_v, item_type, nullptr))));
+        loop_body.push_back(al, list_append);
+        list_concat_stmt = ASRUtils::STMT(ASR::make_DoLoop_t(
+            al, loc, do_loop_head, loop_body.p, loop_body.n));
+        list_while_loop = PassUtils::replace_doloop(
+            al, *ASR::down_cast<ASR::DoLoop_t>(list_concat_stmt));
+        for( size_t i = 0; i < list_while_loop.size(); i++ ) {
+            body.push_back(al, list_while_loop[i]);
+        }
+
+        // Return
+        list_concat_stmt = ASRUtils::STMT(ASR::make_Return_t(al, loc));
+        body.push_back(al, list_concat_stmt);
+
+        ASR::asr_t *fn = ASR::make_Function_t(
+            al, loc,
+            /* a_symtab */ list_concat_symtab,
+            /* a_name */ s2c(al, fn_name),
+            /* a_args */ arg_exprs.p,
+            /* n_args */ arg_exprs.n,
+            /* a_body */ body.p,
+            /* n_body */ body.n,
+            /* a_return_var */ arg_expr,
+            ASR::abiType::Source,
+            ASR::accessType::Public, ASR::deftypeType::Implementation,
```

Comment:
Is this correct?

Reply:
Okay. Wait. So for every list concatenation operation the, `create_list_concat` function is called and hence everytime `global_scope->add_symbol` will be called. I think we need to add it only once. I think you can add a flag in this class and if its false then only execute the body of this execution otherwise just do return. Does that make sense? Or you already handle this?

Reply:
Yup, it make sense. Sorry, I totally forgot it.
Also, instead of using flag. Why not check the symtab whether the function is created. If so, we pass or else create one. Is it possible to do that?

Reply:
So here are the cases to consider,

1. `_lcompilers_list_concat` is already defined by the user and is present in the `global_scope`. In this case `global_scope->resolve_symbol` will always give non-`nullptr` result. It will be hard to distinguish between user's function and what we want to add.
2. `_lcompilers_list_concat` is not defined by the user. In this case it is possible to just query the `global_scope` and get the correct result.

So if we receive non-`nullptr` result we never know whether its user's function or our own function. 

Though note that I am not sure if its even possible for the users to add a function to global scope. But I would play safe since if its not possible for the users to add functions to global scope, it might become possible in future and then it will fail.

Reply:
I agree.

Reply:
For every `ListConcat`, this class will be initialize. Then, how do I store the flag?

Reply:
I think the flag can be stored in `ListConcatVisitor` and while initialising `ReplaceListConcat` pass it into its constructor. So we won't depend on static variables and initialise the list concat function only once. Make sense?

Reply:
> For every `ListConcat`, this class will be initialize. Then, how do I store the flag?

I don't think. The replacer is initialised only once along with list concat visitor. Anyways its still possible to use the flag approach, just keep at the right level of control flow.

Reply:
I applied your suggestions